File: assets/js/analyses-affiliations/your-seat.js
/* assets/js/analyses-affiliations/your-seat.js — the People capstone: the network read from one chair */
(function () {
  "use strict";
  (window.AnalysesRegistry = window.AnalysesRegistry || {
    _q: [], register: function (s, d) { this._q.push([s, d]); }
  }).register("your-seat", {
    prose: {
      intro:
        "<p>Every panel so far reads the whole network at once — all the rooms, all the waves, " +
        "the full transition table. This one turns the map around and reads it from one chair: " +
        "pick any name (or find yours) and the six analyses above are re-asked about that single " +
        "person — their strongest threads, their open doors, their chapters, what history " +
        "suggests comes next. Seats are deep-linkable: <code>?p=</code> in the URL carries a " +
        "seat, so anyone can be handed theirs.</p>",
      how:
        "<p>Nothing new is computed here — this panel is a gather operation, not a model. Each " +
        "of the six analyses already published its results as data: the projection&rsquo;s " +
        "threads, the shared-rooms-without-a-paper pairs, the cohort spans, the transition " +
        "table, the range scores, the solo organizations. This panel indexes all six by one " +
        "person id and lays the rows side by side — like reading a single row of every " +
        "attention map instead of staring at six full matrices.</p>" +
        "<p>The only judgment calls are editorial: which five threads to surface, and reading " +
        "&ldquo;what&rsquo;s next&rdquo; as the two most likely moves out of the person&rsquo;s " +
        "current kind of room, taken straight from the pipeline panel&rsquo;s transition row. " +
        "Because it is a view and not a recomputation, every number here can be checked against " +
        "the panel it came from.</p>",
      method:
        "<p>Ego-network (ego-centric) analysis: fix a focal node, then describe its alters, " +
        "ties, and attributes — the standard treatment is Wasserman &amp; Faust, <em>Social " +
        "Network Analysis: Methods and Applications</em>, Cambridge University Press, 1994. " +
        "All numbers are read verbatim from the six parent panels&rsquo; published data; " +
        "nothing is re-derived. Tie weights come from the career map&rsquo;s " +
        "nesting-discounted bipartite projection (a shared lab displaces its university in the " +
        "weight), so threads here match the map page exactly. Caveats inherit from the " +
        "parents: the sweep records public bios only; &ldquo;open doors&rdquo; count rooms " +
        "both people verifiably shared without a joint paper; and three people&rsquo;s " +
        "chapters carry no dates, so their pipeline reading is omitted rather than guessed.</p>"
    },
    render: function (el, data, shared) {
      var d = data.data, C = shared.colors;
      // org-type palette — keep in sync across the analyses-affiliations panels + the map page
      var TYPE = { lab: "#8c510a", program: "#2166ac", company: "#7b3294",
                   community: "#35978f", university: "#98917f" };

      var totalW = el.clientWidth || 680;
      var miniW = 200, gap = 24;
      var wide = totalW >= 660;
      var W = Math.max(360, Math.min(wide ? totalW - miniW - gap : totalW, 720));
      var twoCol = W >= 540;

      var root = d3.select(el);
      var header = root.append("div")
        .style("display", "flex").style("flex-wrap", "wrap")
        .style("justify-content", "space-between").style("align-items", "flex-end")
        .style("gap", "10px 16px").style("margin-bottom", "14px")
        .style("max-width", (wide ? W + gap + miniW : W) + "px");
      var hLeft = header.append("div");
      var hRight = header.append("div").style("flex", "0 0 auto");

      var body = root.append("div")
        .style("display", "flex").style("flex-wrap", "wrap")
        .style("gap", gap + "px").style("align-items", "flex-start");
      var left = body.append("div").style("flex", "0 0 " + W + "px").style("max-width", "100%");
      var grid = left.append("div")
        .style("display", "grid")
        .style("grid-template-columns", twoCol ? "repeat(2, minmax(0, 1fr))" : "1fr")
        .style("column-gap", "26px").style("align-items", "start");
      var right = body.append("div").style("flex", "0 0 " + miniW + "px");

      // ---- resolve the seat: ?p= → localStorage identity → the host's chair ----
      var qp = new URLSearchParams(location.search).get("p");
      var cur = null, unknown = null;
      if (qp) {
        if (d.people[qp]) cur = qp; else unknown = qp;
      } else {
        try {
          var idn = JSON.parse(localStorage.getItem("network_identity") || "null");
          if (idn && idn.id && d.people[idn.id]) cur = idn.id;
        } catch (e) { /* malformed identity — fall through */ }
        if (!cur) cur = "alex loftus";
      }

      // ---- person picker (chrome — the one place this panel says "you") ----
      var byLabel = {};
      d.order.forEach(function (id) { byLabel[shared.labelOf(id).toLowerCase()] = id; });
      var input = hRight.append("input")
        .attr("list", "your-seat-names")
        .attr("placeholder", "pick a name — or find yours")
        .style("font-size", "12.5px").style("color", C.ink)
        .style("background", "#fff").style("border", "1px solid " + C.hair)
        .style("border-radius", "4px").style("padding", "5px 9px").style("width", "190px");
      var dl = hRight.append("datalist").attr("id", "your-seat-names");
      d.order.forEach(function (id) {
        dl.append("option").attr("value", shared.labelOf(id));
      });
      input.on("input change", function () {
        var id = byLabel[this.value.trim().toLowerCase()];
        if (!id || id === cur) return;
        cur = id; unknown = null;
        var u = new URL(location.href);
        u.searchParams.set("p", id);
        history.replaceState(null, "", u.toString());   // keeps the #your-seat hash
        this.value = "";
        drawPerson(id);
      });

      // ---- mini-map: rebuilt per seat so the ego's threads wear their colors ----
      var mini = { highlight: function () {} };
      function drawMap(pid, capText) {
        right.html("");
        var tieSet = {};
        if (pid) {
          d.people[pid].tie_ids.forEach(function (id) { tieSet[id] = true; });
          tieSet[pid] = true;
        }
        mini = shared.minimap(right.node(), function (id) {
          return tieSet[id] ? shared.colorOf(id) : null;
        }, pid ? { onReady: function (api) { api.highlight(pid); } } : {});
        right.append("div").attr("class", "m-map-cap").style("text-align", "left")
          .text(capText);
      }

      // ---- shared row plumbing ----
      function hoverPerson(row, otherId, tipFn) {
        row.style("cursor", "default")
          .on("mouseenter", function (evt) {
            d3.select(this).style("background", "#f4f1e9");
            mini.highlight(otherId);
            shared.tooltip.show(tipFn(), evt);
          })
          .on("mousemove", function (evt) { shared.tooltip.show(tipFn(), evt); })
          .on("mouseleave", function () {
            d3.select(this).style("background", null);
            mini.highlight(cur);              // restore the seat's own highlight
            shared.tooltip.hide();
          });
      }
      function orgHtml(names) {            // org names in their (muted) type colors
        return names.map(function (o) {
          return "<span style='color:" + (TYPE[d.org_types[o]] || C.muted) + "'>" +
            shared.esc(o) + "</span>";
        }).join("<span style='color:" + C.muted + "'> · </span>");
      }
      function card(title) {
        var c = grid.append("div").style("min-width", "0")
          .style("border-top", "1px solid " + C.hair).style("padding", "10px 0 16px");
        c.append("div").attr("class", "m-axis").style("color", C.muted)
          .style("font-weight", "600").style("letter-spacing", ".07em")
          .style("text-transform", "uppercase").style("margin-bottom", "7px")
          .text(title);
        return c;
      }
      function note(c, txt) {
        c.append("div").style("font-size", "12.5px").style("color", C.muted)
          .style("line-height", "1.45").text(txt);
      }
      function spanText(ch) {
        if (ch.start == null && ch.end == null) return "undated";
        if (ch.spell) return String(ch.start);
        if (ch.ongoing) return ch.start + "–";
        if (ch.start == null) return "–" + ch.end;
        return ch.start + "–" + ch.end;
      }

      // ---- the six cards ----
      function threadsCard(p, label, title) {
        var c = card(title);
        if (!p.ties.length) {
          note(c, "no threads in the projection yet — " + label +
            "’s rooms don’t overlap the chat’s shared rooms.");
          return;
        }
        var maxW = d3.max(p.ties, function (t) { return t.w; });
        p.ties.forEach(function (t) {
          var row = c.append("div").style("padding", "4px 2px");
          row.append("div").style("font-size", "12.5px").style("color", C.ink)
            .html(shared.seatLink(t.other));
          row.append("div").style("height", "3px")
            .style("width", Math.max(6, Math.round(100 * t.w / maxW)) + "%")
            .style("background", shared.colorOf(t.other)).style("opacity", 0.85)
            .style("margin", "3px 0 2px");
          row.append("div").style("font-size", "11px").style("line-height", "1.5")
            .html(orgHtml(t.orgs));
          hoverPerson(row, t.other, function () {
            return "<strong>" + shared.esc(shared.labelOf(t.other)) + "</strong>" +
              "<div style='color:" + C.muted + "'>thread weight " + shared.fmt.num(t.w) +
              "</div><div style='margin-top:2px'>" + orgHtml(t.orgs) + "</div>";
          });
        });
        if (p.n_ties > p.ties.length) {
          c.append("div").style("font-size", "11px").style("color", C.muted)
            .style("margin-top", "4px")
            .text("the 5 strongest of " + p.n_ties + " threads — all of them lit on the map");
        }
      }

      function invitationsCard(p, label, promoted) {
        var c = card("Open invitations");
        if (!p.invitations.length) {
          note(c, "every shared room on " + label + "’s record has already produced a paper.");
          return;
        }
        if (promoted) {
          c.append("div").style("font-size", "12.5px").style("color", "#4a463f")
            .style("line-height", "1.45").style("margin-bottom", "6px")
            .html("the network is holding <strong>" + p.n_invitations +
              "</strong> open door" + (p.n_invitations === 1 ? "" : "s") + " for " +
              shared.esc(label) + ".");
        }
        p.invitations.forEach(function (v) {
          var row = c.append("div").style("font-size", "12.5px").style("padding", "3px 2px")
            .style("line-height", "1.45");
          row.html("<span style='color:" + C.ink + "'>" +
            shared.seatLink(v.other) + "</span><span style='color:" + C.muted +
            "'> — shared " + v.orgs.map(shared.esc).join(", ") + "</span>");
          hoverPerson(row, v.other, function () {
            return "<strong>" + shared.esc(shared.labelOf(v.other)) + "</strong>" +
              "<div style='color:" + C.muted + "'>" + v.overlap + " shared room" +
              (v.overlap === 1 ? "" : "s") + ", no joint paper yet</div>" +
              "<div style='margin-top:2px'>" + orgHtml(v.orgs) + "</div>";
          });
        });
        if (p.n_invitations > p.invitations.length) {
          c.append("div").style("font-size", "11px").style("color", C.muted)
            .style("margin-top", "4px")
            .text("and " + (p.n_invitations - p.invitations.length) + " more open doors");
        }
      }

      function chaptersCard(p, label) {
        var c = card("Chapters");
        if (!p.chapters.length) {
          note(c, "no chapters on record for " + label + " yet.");
          return;
        }
        p.chapters.forEach(function (ch) {
          var row = c.append("div").style("display", "flex").style("align-items", "baseline")
            .style("gap", "7px").style("padding", "3px 2px").style("font-size", "12.5px");
          row.append("span").style("width", "9px").style("height", "9px")
            .style("background", TYPE[ch.type] || C.other)
            .style("flex", "0 0 auto").style("align-self", "center");
          row.append("span").style("color", C.ink).style("flex", "1 1 auto")
            .style("min-width", "0").text(ch.org);
          row.append("span").attr("class", "m-axis").style("color", C.muted)
            .style("flex", "0 0 auto").text(spanText(ch));
          var chTip = function () {
            return "<strong>" + shared.esc(ch.org) + "</strong>" +
              "<div style='color:" + C.muted + "'>" + shared.esc(ch.type) + " · " +
              shared.esc(spanText(ch)) + "</div>" +
              (ch.role ? "<div style='margin-top:2px'>" + shared.esc(ch.role) + "</div>" : "");
          };
          row.style("cursor", "default")
            .on("mouseenter", function (evt) {
              d3.select(this).style("background", "#f4f1e9");
              shared.tooltip.show(chTip(), evt);
            })
            .on("mousemove", function (evt) { shared.tooltip.show(chTip(), evt); })
            .on("mouseleave", function () {
              d3.select(this).style("background", null);
              shared.tooltip.hide();
            });
        });
      }

      function pipelineCard(p, label) {
        var c = card("The pipeline says");
        if (!p.pipeline) {
          note(c, label + "’s chapters carry no dates yet, so the pipeline has nothing to read.");
          return;
        }
        var tape = c.append("div").style("display", "flex").style("flex-wrap", "wrap")
          .style("gap", "3px").style("align-items", "center");
        p.pipeline.seq.forEach(function (t, i) {
          var tkTip = function () {
            return "dated chapter " + (i + 1) + " of " + p.pipeline.seq.length +
              " — <span style='color:" + TYPE[t] + "'>" + shared.esc(t) + "</span>";
          };
          tape.append("span").style("width", "10px").style("height", "10px")
            .style("background", TYPE[t]).style("display", "inline-block")
            .style("flex", "0 0 auto").style("cursor", "default")
            .on("mouseenter", function (evt) { shared.tooltip.show(tkTip(), evt); })
            .on("mousemove", function (evt) { shared.tooltip.show(tkTip(), evt); })
            .on("mouseleave", function () { shared.tooltip.hide(); });
        });
        var curT = p.pipeline.current, nx = p.pipeline.next;
        c.append("div").style("font-size", "12.5px").style("margin-top", "7px")
          .html("<span style='color:" + C.muted + "'>now in a</span> <span style='color:" +
            TYPE[curT] + "'>" + shared.esc(curT) + "</span>");
        c.append("div").style("font-size", "12.5px").style("color", C.muted)
          .style("margin-top", "3px").style("line-height", "1.45")
          .html("history suggests next: <span style='color:" + TYPE[nx[0].type] + "'>" +
            shared.esc(nx[0].type) + "</span> (" + shared.fmt.pct(nx[0].p) +
            "), then <span style='color:" + TYPE[nx[1].type] + "'>" + shared.esc(nx[1].type) +
            "</span> (" + shared.fmt.pct(nx[1].p) + ")");
      }

      function rangeCard(p, label) {
        var c = card("Range");
        var r = p.range;
        var line = c.append("div").style("font-size", "12.5px").style("color", "#4a463f")
          .style("line-height", "1.5");
        if (r.low) {
          line.html(shared.esc(label) + "’s range is still being written — <strong>" + r.n +
            "</strong> room" + (r.n === 1 ? "" : "s") + " on record so far.");
        } else {
          line.html(shared.esc(label) + "’s record covers <strong>" + r.n +
            "</strong> rooms — effectively <strong>" + shared.fmt.num(r.n1) +
            "</strong> different kinds of institution.");
        }
        if (r.quadrant) {
          c.append("div").style("font-size", "12.5px").style("font-style", "italic")
            .style("color", "#6b665d").style("margin-top", "4px")
            .text("a " + r.quadrant + ".");
        }
      }

      function embassiesCard(p, label) {
        var c = card("Embassies");
        if (!p.embassies.n) {
          note(c, "every organization on " + label + "’s record is shared with someone here — " +
            "no solo outposts, all company.");
          return;
        }
        c.append("div").style("font-size", "12.5px").style("color", "#4a463f")
          .style("line-height", "1.45")
          .html("<strong>" + p.embassies.n + "</strong> room" +
            (p.embassies.n === 1 ? "" : "s") + " where " + shared.esc(label) +
            " is the chat’s only member:");
        c.append("div").style("font-size", "11.5px").style("margin-top", "5px")
          .style("line-height", "1.7").html(orgHtml(p.embassies.orgs));
        if (p.embassies.n > p.embassies.orgs.length) {
          c.append("div").style("font-size", "11px").style("color", C.muted)
            .text("and " + (p.embassies.n - p.embassies.orgs.length) + " more.");
        }
      }

      // ---- per-seat assembly ----
      function drawHeader(title, metaHtml) {
        hLeft.html("");
        hLeft.append("h3").attr("class", "m-label")
          .style("margin", "0").style("font-size", "17px").style("font-weight", "600")
          .style("color", C.ink).text(title);
        hLeft.append("div").style("font-size", "12.5px").style("color", C.muted)
          .style("margin-top", "2px").html(metaHtml);
      }

      function drawPerson(pid) {
        var p = d.people[pid], label = shared.labelOf(pid);
        var bits = [];
        if (p.now.city) bits.push(shared.esc(p.now.city));
        bits.push(p.now.no_current ? "between rooms" : shared.esc(p.now.org || ""));
        drawHeader(label + "’s seat", bits.filter(Boolean).join(" · "));
        drawMap(pid, label + "’s corner of the map — their threads wear community colors; " +
          "hover a thread or an invitation to place its person");

        grid.html("");
        if (p.sparse) {            // no ties or invitations yet — invite, but show what IS here
          grid.append("div").style("grid-column", "1 / -1")
            .style("border-top", "1px solid " + C.hair).style("padding", "10px 2px 4px")
            .style("font-size", "12.5px").style("color", "#6b665d").style("line-height", "1.5")
            .html("<strong>The map is young in this corner</strong> — no shared rooms on record " +
              "for " + shared.esc(label) + " yet. " +
              '<a href="/networks/affiliations/?edit=1&p=' + encodeURIComponent(pid) +
              '" style="color:#6b665d">add a chapter on the career map →</a>');
        } else if (p.n_ties + p.n_invitations <= 1 && p.chapters.length <= 1) {
          var nroom = p.chapters.length;                 // thin-but-not-empty: a new arrival
          grid.append("div").style("grid-column", "1 / -1")
            .style("border-top", "1px solid " + C.hair).style("padding", "10px 2px 4px")
            .style("font-size", "12.5px").style("color", "#6b665d").style("line-height", "1.5")
            .html("<strong>The map is young in this corner</strong> — " + shared.esc(label) +
              " has " + (nroom ? "just one room" : "no rooms") + " on record so far, so the " +
              "threads here are still thin. This seat fills in as more of their affiliations are " +
              "added. " +
              '<a href="/networks/affiliations/?edit=1&p=' + encodeURIComponent(pid) +
              '" style="color:#6b665d">add a chapter on the career map →</a>');
        }
        if (p.n_ties < 3) {            // a young corner: lead with the open doors
          invitationsCard(p, label, true);
          threadsCard(p, label, "Newest threads");
        } else {
          threadsCard(p, label, "Strongest threads");
          invitationsCard(p, label, false);
        }
        chaptersCard(p, label);
        pipelineCard(p, label);
        rangeCard(p, label);
        embassiesCard(p, label);
      }

      function drawUnknown(raw) {
        drawHeader("This seat is being set",
          "“" + shared.esc(raw) + "” isn’t on the map yet");
        drawMap(null, "the map, holding the room until the seat is set");
        grid.html("");
        grid.append("div").style("grid-column", "1 / -1")
          .style("border-top", "1px solid " + C.hair).style("padding", "18px 2px")
          .style("font-size", "13.5px").style("color", "#4a463f")
          .style("line-height", "1.55").style("max-width", "56ch")
          .html("This seat is being set — new members get their seat after the next data refresh — usually a day or two. " +
            'Until then, the <a href="/networks/affiliations/" style="color:#6b665d">' +
            "career map</a> holds everyone whose chapters are already on record.");
      }

      if (unknown != null) drawUnknown(unknown); else drawPerson(cur);
    }
  });
})();

File: experiments/coauthorship/analyses-affiliations/your-seat.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""your-seat — the People capstone: a per-person digest gathered from the six panels.
Run: cd experiments/coauthorship/analyses-affiliations && uv run your-seat.py"""
import json
import re
from collections import defaultdict
from pathlib import Path

REPO = Path(__file__).resolve().parents[3]
assert (REPO / "assets").exists(), f"REPO mis-resolved: {REPO}"

AFF = json.loads((REPO / "assets/data/affiliations.json").read_text())
PANELS = REPO / "assets/data/analyses-affiliations"
SRP = json.loads((PANELS / "same-rooms-no-paper.json").read_text())["data"]
ERAS = json.loads((PANELS / "eras.json").read_text())["data"]
PIPE = json.loads((PANELS / "the-pipeline.json").read_text())["data"]
RANGE = json.loads((PANELS / "range.json").read_text())["data"]
EMB = json.loads((PANELS / "embassies.json").read_text())["data"]
WWN = json.loads((PANELS / "where-we-are-now.json").read_text())["data"]
SHARED = json.loads((PANELS / "shared.json").read_text())
OUT = PANELS / "your-seat.json"

NODES = set(SHARED["nodes"])
people = {p["id"]: p for p in AFF["people"]}
assert len(people) >= 48, len(people)
assert set(people) <= NODES
assert len({p["label"] for p in people.values()}) == len(people)  # labels unique → picker can key on them
for pid, p in people.items():
    assert p["label"] == SHARED["nodes"][pid]["label"], pid
assert "alex loftus" in people  # the JS default seat

orgs = {o["id"]: o for o in AFF["orgs"]}
assert len({o["label"] for o in orgs.values()}) == len(orgs)
org_label = {oid: o["label"] for oid, o in orgs.items()}
# label -> type, for the JS type-coloring of org names (ties + embassies ship labels only)
org_types = {o["label"]: o["type"] for o in sorted(orgs.values(), key=lambda o: o["label"])}

# ---- year parsing — same start/end regex spec as eras.py (kept in sync by hand) ----
NOW = 2026


def start_year(y):
    m = re.match(r"^\s*(\d{4})", y or "")
    return int(m.group(1)) if m else None


def parse_end(y):
    ys = re.findall(r"\d{4}", y or "")
    if not ys:
        return None
    return NOW if re.search(r"[–-]\s*$", y) else int(ys[-1])


# ---- ties: the nesting-discounted projection, read per person ------------------------------
ties_by: dict[str, list[dict]] = defaultdict(list)
for e in AFF["projection"]:
    assert e["a"] in NODES and e["b"] in NODES, (e["a"], e["b"])
    shared_labels = [org_label[s] for s in e["shared"]]
    for me, other in ((e["a"], e["b"]), (e["b"], e["a"])):
        ties_by[me].append({"other": other, "w": e["weight"], "orgs": shared_labels})
for ls in ties_by.values():
    ls.sort(key=lambda t: (-t["w"], t["other"]))

# ---- invitations: kind=="open" pairs from same-rooms-no-paper ------------------------------
inv_by: dict[str, list[dict]] = defaultdict(list)
for pr in SRP["pairs"]:
    if pr["kind"] != "open":
        continue
    assert pr["a"] in NODES and pr["b"] in NODES
    assert all(o in org_types for o in pr["orgs"]), pr["orgs"]
    for me, other in ((pr["a"], pr["b"]), (pr["b"], pr["a"])):
        inv_by[me].append({"other": other, "orgs": pr["orgs"], "overlap": pr["overlap"]})
for ls in inv_by.values():
    ls.sort(key=lambda v: (-v["overlap"], v["other"]))
for pid in people:  # published per-person open counts must agree with the pair list
    assert SRP["open_count_by_person"].get(pid, 0) == len(inv_by.get(pid, [])), pid

# ---- chapters: eras spans where drawn, supplemented from raw links everywhere else ---------
era_member = {}  # (pid, oid) -> (org record, member record)
for o in ERAS["orgs"]:
    for m in o["members"]:
        era_member[(m["id"], o["id"])] = (o, m)

links_by_person: dict[str, list[dict]] = defaultdict(list)
seen = set()
for l in AFF["links"]:
    key = (l["person"], l["org"])
    assert key not in seen, key  # one chapter per (person, org)
    seen.add(key)
    assert l["org"] in orgs and l["person"] in people
    links_by_person[l["person"]].append(l)
assert all(k in seen for k in era_member), "eras member without a backing link"

chapters_of: dict[str, list[dict]] = {}
for pid in sorted(people):
    ch = []
    for l in links_by_person[pid]:
        oid = l["org"]
        if (pid, oid) in era_member:  # drawn org: reuse the eras record verbatim
            o, m = era_member[(pid, oid)]
            ch.append({"org": o["label"], "type": o["type"], "start": m["start"],
                       "end": m["end"], "ongoing": m["ongoing"], "spell": m["spell"],
                       "role": m["role"]})
        else:  # < 3-member org, or undated member of a drawn org
            years = l.get("years") or ""
            s, e = start_year(years), parse_end(years)
            ch.append({"org": org_label[oid], "type": orgs[oid]["type"], "start": s,
                       "end": e, "ongoing": bool(re.search(r"[–-]\s*$", years)),
                       "spell": s is not None and s == e, "role": l.get("role") or ""})
    ch.sort(key=lambda c: (c["start"] if c["start"] is not None else 9999,
                           c["end"] if c["end"] is not None else 9999, c["org"]))
    chapters_of[pid] = ch
assert sum(map(len, chapters_of.values())) == len(AFF["links"]) == 237

# ---- pipeline: dated type sequence in career order; next from the published probs row ------
TYPES = PIPE["types"]
assert len(TYPES) == 5 and all(len(r) == 5 for r in PIPE["probs"])
dated_by: dict[str, list[dict]] = defaultdict(list)
for l in AFF["links"]:
    if start_year(l.get("years")) is not None:
        dated_by[l["person"]].append(l)
for ls in dated_by.values():  # same career order as the-pipeline.py: (start year, org id)
    ls.sort(key=lambda l: (start_year(l["years"]), l["org"]))
assert set(dated_by) == set(PIPE["latest_type"])  # 45 people; 3 have no dated chapters


def next_two(cur: str) -> list[dict]:
    row = PIPE["probs"][TYPES.index(cur)]
    ranked = sorted(zip(TYPES, row), key=lambda tp: (-tp[1], TYPES.index(tp[0])))
    return [{"type": t, "p": p} for t, p in ranked[:2]]


# ---- range: published scores + the quadrant word -------------------------------------------
rng_by = {p["id"]: p for p in RANGE["people"]}
assert set(rng_by) == set(people)


def quadrant(r: dict) -> str | None:
    if r["n1"] >= 3.5:
        return "wide traveler"
    if r["n"] <= 4 and r["n1"] <= 1.5 and not r["low"]:
        return "deep anchor"
    return None


# ---- embassies: solo-org labels per holder --------------------------------------------------
solo_rows = [t for t in EMB["tally"] if t["n"] == 1]
assert len(solo_rows) == 1 and len(solo_rows[0]["orgs"]) == EMB["n_solo"] == 125
solo_by: dict[str, list[str]] = defaultdict(list)
for o in solo_rows[0]["orgs"]:
    assert len(o["members"]) == 1 and o["label"] in org_types
    solo_by[o["members"][0]].append(o["label"])
for ls in solo_by.values():
    ls.sort()
assert {p: len(v) for p, v in solo_by.items()} == EMB["holders"]

# ---- now: the census row -------------------------------------------------------------------
wwn_by = {p["id"]: p for p in WWN["people"]}
assert set(wwn_by) == set(people)
no_cur = set(WWN["no_current_ids"])
assert all((not wwn_by[pid]["all_current"]) == (pid in no_cur) for pid in people)

# ---- assemble digests ------------------------------------------------------------------------
digests: dict[str, dict] = {}
for pid in sorted(people):
    t_all = ties_by.get(pid, [])
    inv_all = inv_by.get(pid, [])
    w, r = wwn_by[pid], rng_by[pid]
    solo = solo_by.get(pid, [])
    dig = {
        "ties": t_all[:5],
        "n_ties": len(t_all),
        "tie_ids": sorted(t["other"] for t in t_all),  # full ego set for the minimap
        "invitations": inv_all[:6],
        "n_invitations": len(inv_all),
        "chapters": chapters_of[pid],
        "range": {"n": r["n"], "n1": r["n1"], "low": r["low"], "quadrant": quadrant(r)},
        "embassies": {"orgs": solo[:8], "n": len(solo)},
        "now": {"org": w["all_current"][0]["org"] if w["all_current"] else None,
                "city": w["city"] or None,
                "no_current": pid in no_cur},
        "sparse": not t_all and not inv_all,
    }
    if pid in dated_by:  # omitted entirely when no dated chapters exist
        seq = [orgs[l["org"]]["type"] for l in dated_by[pid]]
        assert seq[-1] == PIPE["latest_type"][pid], pid
        dig["pipeline"] = {"seq": seq, "current": seq[-1], "next": next_two(seq[-1])}
    digests[pid] = dig

sparse_ids = sorted(pid for pid, dd in digests.items() if dd["sparse"])
shipped_others = {t["other"] for dd in digests.values() for t in dd["ties"]} | \
    {v["other"] for dd in digests.values() for v in dd["invitations"]} | \
    {i for dd in digests.values() for i in dd["tie_ids"]}
assert shipped_others <= NODES

order = sorted(people, key=lambda pid: (people[pid]["label"], pid))
assert len(order) == len(people)

headline = (f"There are <strong>{len(people)}</strong> seats at this table — "
            "pick a name and see what the network holds for them.")

payload = {
    "slug": "your-seat",
    "title": "Your seat",
    "headline": headline,
    "data": {"order": order, "people": digests, "org_types": org_types},
}
blob = json.dumps(payload, separators=(",", ":"))
assert len(blob) < 300_000, len(blob)
OUT.write_text(blob)
print(f"[your-seat] OK {len(blob)/1024:.0f}KB — " + re.sub(r"</?strong>", "", headline))
